Deploying a node, installing drand. (#55)
drand/website: docs/operator/deploy.md

# Deployment guide

> This document explains the workflow to have a working group of drand nodes generate randomness. There are three main sections to this guide:

Reviewer: "There are three main sections to this guide", but 4 bullets below

Author: Done!

> - generate the long-term key pairs and the group file.
> - start the daemons.
> - have each node collectively particiate in the distributed key generation (DKG).
> - genearte randomness.

Reviewer: ```suggestion
1. Generate the long-term key pairs and the group file.
1. Start the daemons.
1. Have each node collectively participate in the distributed key generation (DKG).
1. Generate randomness.
```

## Setup

The setup process for a drand node consists of the following steps:

1. Generate the long-term key pair for each node.
1. Start the drand daemon on each node.
1. The leader starts the command as a coordinator & every participant connects to the coordinator to setup the network.

> This document explains how to do the setup with the drand binary itself. If you want to install drand using Docker, follow the [Docker instructions instead](/operate/docker).

Reviewer: ```suggestion
This document explains how to do the setup with the drand binary itself. If you want to install drand using Docker, follow the [Docker instructions instead](/operate/docker/).
```

### Long-term key

> Each drand node needs a public and secret key to interact with the rest of the network. To generate these keys run `drand generate-keypair` followed by the address of your node:

Reviewer: Link to CLI docs for `generate-keypair` please!

Author: Done!

```bash
drand generate-keypair drand.example.com
```

> The address must be reachable over a TLS connection directly, or via a reverse proxy setup. If you need a non-secured channel, you can pass the `--tls-disable` flag, although this is not recommended. Disabling TLS should only really be done when running a [local deployment](/operate/local-deploy).

Reviewer: ```suggestion
The address must be reachable over a TLS connection directly, or via a reverse proxy setup. If you need a non-secured channel, you can pass the `--tls-disable` flag, although this is not recommended. Disabling TLS should only really be done when running a [local deployment](/operate/local-deploy/).
```

> The default location for your keys is `/home/<USERNAME>/.drand`. You can specify where you want the keys to be saved by using the `--folder` flag:

Reviewer: I'm pretty sure it's the system home directory in which case it's not always `/home/<USERNAME>`:

```suggestion
The default location for your keys is `~/.drand`. You can specify where you want the keys to be saved by using the `--folder` flag:
```

##### TLS setup: Nginx with Let's Encrypt

Running drand behind a reverse proxy is the **default** method of deploying drand. Such a setup greatly simplifies TLS management issues (renewal of certificates, etc.). We provide here the minimum setup using [Nginx](https://www.nginx.com/) and [certbot](https://certbot.eff.org/instructions/) - make sure you have both binaries installed with the latest version; Nginx version must be at least >= `1.13.10` for gRPC compatibility.

1. First, add an entry in the Nginx configuration for drand:

    ```bash
    # /etc/nginx/sites-available/default
    server {
    server_name drand0.example.com;
    listen 443 ssl http2;

    location / {
        grpc_pass grpc://localhost:8080;
        grpc_set_header X-Real-IP $remote_addr;
    }

    location /public/ {
        proxy_pass http://localhost:4444;
        proxy_set_header Host $host;
    }
    location /info {
        proxy_pass http://localhost:4444;
        proxy_set_header Host $host;
    }
    # Add SSL certificates by running certbot --nginx
    }
    ```

    You can change:

>     -. the port on which you want drand to be accessible by changing the line `listen 443 ssl http2` to use any port.
>     -. the port on which the drand binary will listen locally by changing the line `grpc_pass grpc://localhost:8080;` to the private API port and `proxy_pass http://localhost:8080;` to the public API port

Reviewer: ```suggestion
    - the port on which you want drand to be accessible by changing the line `listen 443 ssl http2` to use any port.
    - the port on which the drand binary will listen locally by changing the line `grpc_pass grpc://localhost:8080;` to the private API port and `proxy_pass http://localhost:8080;` to the public API port.
```

drand share --leader --nodes 10 --threshold 6 --secret mysecret901234567890123456789012 --period 30s
```

> ** rest of participants**: Once the coordinator has run the previous command, the rest of the participants must run the following command:

Reviewer: ```suggestion
**Rest of participants**: Once the coordinator has run the previous command, the rest of the participants must run the following command:
```

```bash
drand share --connect <leaderaddress> --secret mysecret901234567890123456789012
```

The flags usage is as follow:

| Flag | Description |
| ---- | ----------- |
| `--leader` | _This_ node is the group coordinator. |
| `--nodes` | The number of nodes in this group. |
| `--threshold` | The minimum number of nodes that need to be online for the network to be live. |
| `--period` | The period of the randomness beacon to use. It must be a valid duration as parsed by Golang's [`time.ParseDuration`](https://golang.org/pkg/time/#ParseDuration) method. |

> | `--secret` | The password that the leader uses to authenticate nodes that want to participate in the group. This password must be at least 32 characters long.

Reviewer: ```suggestion
| `--secret` | The password that the leader uses to authenticate nodes that want to participate in the group. This password must be at least 32 characters long. |
```

```bash
drand share <group-file> --source <entropy-exec> --user-source-only
```

### Group TOML file

Once the DKG phase is done, each node has both a private share and a group file containing the distributed public key. Using the previous commands, the group file will be written to `group.toml`. That updated group file is needed by drand to securely contact drand nodes on their public interface to gather private or public randomness. To view this file, run `drand show group`. If you want to save the output to a file, add the `--out <file>` flag:

```bash
drand show group --out ~/group-config.toml
```

## Randomness Generation

After a successful setup phase, drand will switch to the randomness generation mode _at the genesis time_ specified in the group file. Each node broadcasts _randomness shares_ at regular intervals. Every new random beacon is linked to the previous one in a chain of randomness. Once a node has collected a threshold of shares in the current round, it computes the public, random value, and stores it in its local instance of [BoltDB](https://github.com/coreos/bbolt).

For third party implementations of randomness beacon verification, you need:

> - the distributed public key generated during the setup phase.
> - the period
> - the genesis time of the chain

Reviewer: ```suggestion
- The distributed public key generated during the setup phase.
- The period.
- The genesis time of the chain.
```

}
```

The "gid" simply indicates which group the data belongs to. It is present for scalar and points on the curve, even though scalars are the same on the three groups of bls12-381. The field is present already to be able to accommodate different curves later on.

### Chain information

> To retrieve the chain information, this node participates to, run:

Reviewer: ```suggestion
To retrieve information about the chain this node participates in, run:
```

```bash
drand show chain-info
```

## Updating drand group

Drand allows for "semi-dynamic" group update with a _resharing_ protocol that offers the following:

> - new nodes can join an existing group and get new shares. Note that, in fact, all nodes get _new_ shares after running the resharing protocol.
> - nodes can leave their current group. It may be necessary for nodes that do not wish to operate drand anymore.
> - nodes can update the threshold associated with their current distributed public key.

Reviewer: ```suggestion
- New nodes can join an existing group and get new shares. Note that, in fact, all nodes get _new_ shares after running the resharing protocol.
- Nodes can leave their current group. It may be necessary for nodes that do not wish to operate drand anymore.
- Nodes can update the threshold associated with their current distributed public key.
```
